### ISSUE TITLE
Invalid `init` Function Code for Passing in Arraybuffer

### ISSUE THREAD

**Issue by cretz** [rustwasm/wasm-bindgen#1418]
Please see the following code: https://github.com/rustwasm/wasm-bindgen/blob/a1c457bd152f24bc51eea14bd9cf0f3913f7e74b/crates/cli-support/src/js/mod.rs#L891-L903

The result of `instantiate` according to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate) is `ResultObject`, not just `Instance` as is assumed here. I have to patch this to access `instance.instance` for it to work right.

**Comment by cretz** [comment]
Actually this appears fixed there and is just invalid in my current wasm-pack. Disregard for now, thanks.

**Comment by cretz** [comment]
Ok, reopening, having trouble with latest on Windows. For the simple wasm from a `cargo build --target wasm32-unknown-unknown`:

    # C:\Users\chad.retz\AppData\Local\.wasm-pack\wasm-bindgen-9a795696ca119c40\wasm-bindgen.exe --version
    wasm-bindgen 0.2.40 (7392df8a3)

    C:\Users\chad.retz\AppData\Local\.wasm-pack\wasm-bindgen-9a795696ca119c40\wasm-bindgen.exe target\wasm32-unknown-unknown\debug\sample.wasm --out-dir pkg --web

(I need pkg web because I need to init the WASM array buffer myself). Here is the generated init function:

```js
function init(module_or_path, maybe_memory) {
    let result;
    const imports = { './imedge_manip': __exports };
    if (module_or_path instanceof URL || typeof module_or_path === 'string' || module_or_path instanceof Request) {

        const response = fetch(module_or_path);
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            result = WebAssembly.instantiateStreaming(response, imports)
            .catch(e => {
                console.warn("`WebAssembly.instantiateStreaming` failed. Assuming this is because your server does not serve wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", e);
                return response
                .then(r => r.arrayBuffer())
                .then(bytes => WebAssembly.instantiate(bytes, imports));
            });
        } else {
            result = response
            .then(r => r.arrayBuffer())
            .then(bytes => WebAssembly.instantiate(bytes, imports));
        }
    } else {

        result = WebAssembly.instantiate(module_or_path, imports)
        .then(instance => {
            return { instance, module: module_or_path };
        });
    }
    return result.then(({instance, module}) => {
        wasm = instance.exports;
        init.__wbindgen_wasm_module = module;

        return wasm;
    });
}
```

Note, this code expects `WebAssembly.instantiate` to return a promise for `instance` but it in fact returns a promise for `ResultObject` of which `instance` is a member. I can't find this part of the init script in the codebase even on that version tag (just released)...so I wonder if the download for wasm-pack is acting funny.

**Comment by alexcrichton** [comment]
Thanks for the report! Definitely looks like a bug and should be fixed in #1419
